lib/common/xml.c

```diff
@@ -1703,108 +1703,108 @@ pcmk__xml_log(int log_level, const char *prefix, const xmlNode *data, int depth,
     g_string_free(buffer, TRUE);
 }
 
-// Log XML portions that have been marked as changed
+/*!
+ * \internal
+ * \brief Log XML portions that have been marked as changed
+ *
+ * \param[in] log_level  Priority at which to log the messages
+ * \param[in] data       XML node to log
+ * \param[in] depth      Current indentation level
+ * \param[in] options    Group of \p xml_log_options flags
+ */
 static void
-log_xml_changes(int log_level, const char *prefix, const xmlNode *data,
-                int depth, int options)
+log_xml_changes(int log_level, const xmlNode *data, int depth, int options)
 {
-    xml_node_private_t *nodepriv;
-    char *prefix_m = NULL;
-    xmlNode *child = NULL;
+    const char *prefix_created = "++";
+    const char *prefix_deleted = "--";
+    const char *prefix_modified = "+ ";
+    const char *prefix_moved = "+~";
+    xml_node_private_t *nodepriv = NULL;
 
     if ((data == NULL) || (log_level == LOG_NEVER)) {
         return;
     }
 
     nodepriv = data->_private;
 
-    prefix_m = strdup(prefix);
-    prefix_m[1] = '+';
-
     if (pcmk_all_flags_set(nodepriv->flags, pcmk__xf_dirty|pcmk__xf_created)) {
-        /* Continue and log full subtree */
-        pcmk__xml_log(log_level, prefix_m, data, depth,
+        // Newly created
+        pcmk__xml_log(log_level, prefix_created, data, depth,
                       options
                       |xml_log_option_open
                       |xml_log_option_close
                       |xml_log_option_children);
+        return;
+    }
 
-    } else if (pcmk_is_set(nodepriv->flags, pcmk__xf_dirty)) {
+    if (pcmk_is_set(nodepriv->flags, pcmk__xf_dirty)) {
+        // Deleted, modified, or moved
```

Comment:
I'm not sure deleted is possible in this block

Reply:
Good catch. It is in a sense, but not at the top level. Deleted **attributes** are handled in a nested block. Deletion at the top level is handled by the caller (`xml_log_changes()`).